### PR TITLE
Remove pending activity cancellations when activity completion occurs

### LINKED ISSUE #724: Activity cancellation causes history error
## Expected Behavior

Should not have internal state errors

## Actual Behavior

Has internal state errors, specifically something like:

    BadRequestCancelActivityAttributes: invalid history builder state for action: add-activitytask-cancel-requested-event, ScheduledID: 99

## Steps to Reproduce the Problem

<details>
<summary>main.go</summary>

```go
package main

import (
	"context"
	"fmt"
	"log"
	"os/exec"
	"runtime/debug"
	"strings"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/worker"
	"go.temporal.io/sdk/workflow"
)

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}

const useTemporalite = true
const taskQueue = "my-task-queue"

func run() error {
	var clientOpts client.Options

	// Create a logger that will break on the error we've been waiting for
	errSeenCh := make(chan struct{})
	clientOpts.Logger = &logger{
		excludeLevels: map[string]bool{"DEBUG": true},
		onMessageContains: map[string]func(level, msg string){
			"BadRequestCancelActivityAttributes": func(level, msg string) {
				debug.PrintStack()
				close(errSeenCh)
			},
		},
	}

	// Use temporalite if desired
	if useTemporalite {
		clientOpts.Namespace = "namespace-" + uuid.NewString()
		clientOpts.HostPort = "localhost:9233"
		cmd := exec.Command("temporalite", "start", "--ephemeral", "--namespace", clientOpts.Namespace, "--port", "9233")
		// cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
		if err := cmd.Start(); err != nil {
			return fmt.Errorf("failed starting temporalite: %w", err)
		}
		defer cmd.Process.Kill()
		time.Sleep(200 * time.Millisecond)
	}

	// Create client
	c, err := client.NewClient(clientOpts)
	if err != nil {
		return fmt.Errorf("failed creating client: %w", err)
	}
	defer c.Close()

	// Start workers
	const workerCount = 10
	log.Printf("Starting %v workers", workerCount)
	for i := 0; i < workerCount; i++ {
		w, err := startWorker(c)
		if err != nil {
			return fmt.Errorf("failed starting worker: %w", err)
		}
		defer w.Stop()
	}
	log.Printf("Started %v workers", workerCount)

	// Run "breakers"
	const breakerCount = 10
	log.Printf("Running %v breakers", breakerCount)
	breakerDoneCh := make(chan error, breakerCount)
	for i := 0; i < breakerCount; i++ {
		go func() { breakerDoneCh <- runBreaker(c) }()
	}

	// Wait for them all to finish or error
	for i := 0; i < breakerCount; i++ {
		select {
		case <-errSeenCh:
			return fmt.Errorf("saw BadRequestCancelActivityAttributes")
		case err := <-breakerDoneCh:
			if err != nil {
				return fmt.Errorf("breaker failed: %w", err)
			}
		}
	}
	log.Printf("Breakers done")
	return nil
}

func startWorker(c client.Client) (worker.Worker, error) {
	w := worker.New(c, taskQueue, worker.Options{
		MaxConcurrentActivityExecutionSize: 2,
	})
	w.RegisterWorkflow(SomeWorkflow)
	w.RegisterWorkflow(DoABunchOfActivitiesWorkflow)
	w.RegisterActivity(NoopActivity)
	return w, w.Start()
}

func runBreaker(c client.Client) error {
	for i := 0; i < 20; i++ {
		_, err := c.ExecuteWorkflow(context.Background(), client.StartWorkflowOptions{TaskQueue: taskQueue}, SomeWorkflow)
		if err != nil {
			return fmt.Errorf("failed starting workflow: %w", err)
		}
		time.Sleep(2 * time.Second)
	}
	return nil
}

func SomeWorkflow(ctx workflow.Context) error {
	childCtx, childCancel := workflow.WithCancel(ctx)
	childCtx = workflow.WithChildOptions(childCtx, workflow.ChildWorkflowOptions{
		WorkflowTaskTimeout: time.Minute,
		WaitForCancellation: true,
	})
	childFut := workflow.ExecuteChildWorkflow(childCtx, DoABunchOfActivitiesWorkflow)

	// Wait for start
	if err := childFut.GetChildWorkflowExecution().Get(ctx, nil); err != nil {
		return err
	}

	// Cancel and wait for child done
	childCancel()
	return childFut.Get(ctx, nil)
}

func DoABunchOfActivitiesWorkflow(ctx workflow.Context) error {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Hour,
		HeartbeatTimeout:    20 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2,
			MaximumInterval:    time.Minute * 10,
			MaximumAttempts:    5,
		},
		WaitForCancellation: true,
	})

	childCtx, cancelHandler := workflow.WithCancel(ctx)
	sel := workflow.NewSelector(ctx)

	const activityCount = 100

	for i := 0; i < activityCount; i++ {
		sel.AddFuture(workflow.ExecuteActivity(childCtx, NoopActivity), func(f workflow.Future) {
			if err := f.Get(ctx, nil); err != nil {
				// Cancel all activities
				cancelHandler()
			}
		})
	}

	for i := 0; i < activityCount; i++ {
		sel.Select(ctx)
	}

	return nil
}

func NoopActivity(aCtx context.Context) error { return nil }

type logger struct {
	// Called _after_ logged
	onMessageContains map[string]func(level, msg string)
	excludeLevels     map[string]bool
}

func (l *logger) log(level, msg string, keyvals ...interface{}) {
	if l.excludeLevels[level] {
		return
	}
	msg = fmt.Sprint(append([]interface{}{msg}, keyvals...))
	log.Print([]interface{}{level, msg})
	for str, callback := range l.onMessageContains {
		if strings.Contains(msg, str) {
			callback(level, msg)
		}
	}
}

func (l *logger) Debug(msg string, keyvals ...interface{}) { l.log("DEBUG", msg, keyvals...) }
func (l *logger) Info(msg string, keyvals ...interface{})  { l.log("INFO", msg, keyvals...) }
func (l *logger) Warn(msg string, keyvals ...interface{})  { l.log("WARN", msg, keyvals...) }
func (l *logger) Error(msg string, keyvals ...interface{}) { l.log("ERROR", msg, keyvals...) }
```

</details>

Run the `main.go` code above with https://github.com/DataDog/temporalite on the `PATH`. It should break fairly quickly.

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

We make sure we remove pending activity cancellations when we receive an activity completion from the server, taking care not to incorrectly alter the event sequence identifier incorrectly.

## Why?

This has some history I'll try to explain here in a quick set of bullets.

* The existing implementation in Go SDK collects commands for cancellation before it processes events that may make the cancellation moot
* Therefore, in a racy situation, you can get an activity completion after you've queued up an activity cancellation and if you sent that cancellation, the server fails
* This was mostly fixed in #504 but the special code that handled fixing the event counter when removing the no-longer-needed pending cancel had an issue for activities that were sent after it (e.g. cleanup activities)
* This was mostly fixed in #625 by removing the activity part of #504. In hindsight we instead should have just fixed the event counter issue and left the rest of it alone
* So this issue reinstates the activity-specific no-longer-needed-pending-activity-cancel-removal from #504, which fails the tests from #625 by default, but also adds new logic to make sure the event counter manipulation works correctly

TL;DR - We mostly fixed, we removed part of that fix due to another error, we re-applied the fix and made changes to also fix that other error

## Checklist

1. Closes #724

2. How was this tested:

This relies on a test from #625 to ensure no regressions. Unfortunately the full regression test is a several-second, server-burdening script (see it in #725). Still contemplating where best to put it.

### CODE REVIEW

**Review by yiminc — APPROVED** [review]

File: internal/internal_decision_state_machine.go [hunk 1]
```diff
@@ -824,6 +846,9 @@ func (h *commandsHelper) setCurrentWorkflowTaskStartedEventID(workflowTaskStarte
 	// execution as those canceled command events will show up *after* the workflow task completed event.
 	h.nextCommandEventID = workflowTaskStartedEventID + 2 + h.commandsCancelledDuringWFCancellation
 	h.commandsCancelledDuringWFCancellation = 0
+	// We must change the counter here so that others who mutate
+	// commandsCancelledDuringWFCancellation know it has since been reset
+	h.nextCommandEventIDResetCounter++
```
[the remark this answers is not in the document]
Comment (cretz) [reply internal/internal_decision_state_machine.go:851]: This field created in #382 and this comment was added where it is used:

> We must also add the number of cancel commands that were spawned during cancellation of the workflow
execution as those canceled command events will show up *after* the workflow task completed event.

File: internal/version.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ package internal
 const (
 	// SDKVersion is a semver (https://semver.org/) that represents the version of this Temporal GoSDK.
 	// Server validates if SDKVersion fits its supported range and rejects request if it doesn't.
-	SDKVersion = "1.13.0"
+	SDKVersion = "1.13.1"
```
Comment (cretz) [inline internal/version.go:33]: This is just a temporary change for those trying this out to confirm they are on the right branch. This will be removed before merge.
